scarb/src/compiler/plugin/proc_macro/host.rs

```diff
@@ -16,22 +22,131 @@ pub struct ProcMacroHostPlugin {
+    MacroCall,
+    /// `#[proc_macro_name]`
+    Attribute,
+    /// `#[derive(...)]`
+    Derive,
+}
+
+#[derive(Debug)]
+pub struct ProcMacroInput {
+    pub id: ProcMacroId,
+    pub kind: ProcMacroKind,
+}
+
 impl ProcMacroHostPlugin {
     pub fn new(macros: Vec<Arc<ProcMacroInstance>>) -> Self {
         Self { macros }
     }
+
+    /// Handle `proc_macro_name!` expression.
+    fn handle_macro(
+        &self,
+        _db: &dyn SyntaxGroup,
+        _item_ast: ast::ModuleItem,
+    ) -> Vec<ProcMacroInput> {
+        // Todo(maciektr): Implement.
+        Vec::new()
+    }
+
+    /// Handle `#[proc_macro_name]` attribute.
+    fn handle_attribute(
+        &self,
+        db: &dyn SyntaxGroup,
+        item_ast: ast::ModuleItem,
+    ) -> Vec<ProcMacroInput> {
+        let attrs = match item_ast {
+            ast::ModuleItem::Struct(struct_ast) => Some(struct_ast.attributes(db)),
+            ast::ModuleItem::Enum(enum_ast) => Some(enum_ast.attributes(db)),
+            ast::ModuleItem::ExternType(extern_type_ast) => Some(extern_type_ast.attributes(db)),
+            ast::ModuleItem::ExternFunction(extern_func_ast) => {
+                Some(extern_func_ast.attributes(db))
+            }
+            ast::ModuleItem::FreeFunction(free_func_ast) => Some(free_func_ast.attributes(db)),
+            _ => None,
+        };
+
+        attrs
+            .map(|attrs| attrs.structurize(db))
+            .unwrap_or_default()
+            .iter()
+            .filter(|attr| self.declared_attributes().contains(&attr.id.to_string()))
+            .map(|attr| ProcMacroInput {
+                id: attr.id.clone(),
+                kind: ProcMacroKind::Attribute,
+            })
+            .collect_vec()
+    }
+
+    /// Handle `#[derive(...)]` attribute.
+    fn handle_derive(
+        &self,
+        _db: &dyn SyntaxGroup,
+        _item_ast: ast::ModuleItem,
+    ) -> Vec<ProcMacroInput> {
+        // Todo(maciektr): Implement.
+        Vec::new()
+    }
 }
 
 impl MacroPlugin for ProcMacroHostPlugin {
     fn generate_code(
         &self,
-        _db: &dyn SyntaxGroup,
-        _item_ast: ModuleItem,
+        db: &dyn SyntaxGroup,
+        item_ast: ast::ModuleItem,
         _metadata: &MacroPluginMetadata<'_>,
     ) -> PluginResult {
         // Apply expansion to `item_ast` where needed.
-        // TODO(maciektr): Implement
-        PluginResult::default()
+        let expansions = self
+            .handle_macro(db, item_ast.clone())
+            .into_iter()
+            .chain(self.handle_attribute(db, item_ast.clone()))
+            .chain(self.handle_derive(db, item_ast.clone()));
+
+        let mut token_stream = TokenStream::from_item_ast(db, item_ast);
+        let mut modified = false;
+        for input in expansions {
+            let instance = self
+                .macros
+                .iter()
+                .find(|m| m.declared_attributes().contains(&input.id.to_string()))
+                .unwrap();
```

Comment:
Multiple macros per instance will be handled later on

Reply:
Yes. The ProcMacroInstance represents a package (Scarb package with Rust source code), that implements the proc macro. For now, it is only possible to have one macro per in such package. Ultimately, it should be possible to define multiple macros in a single package.